File: python/src/tools/learn.py
"""Learn tool — extract and save knowledge from a conversation transcript."""

import sys

from ..extractor.extractor import extract_knowledge
from ..store.metadata import MetadataStore
from ..store.scope import detect_project_root
from ..store.types import KnowledgeScope
from ..store.vector import VectorStore


async def handle_learn(
    transcript_path: str,
    session_id: str,
    scope: str | None = None,
) -> str:
    """Extract and save knowledge from a conversation transcript."""
    project_root = detect_project_root()
    project_name = project_root.split("/")[-1] if project_root else None

    # Extract knowledge from transcript
    chunks = await extract_knowledge(
        transcript_path=transcript_path,
        session_id=session_id,
        trigger="manual",
        project_name=project_name,
        scope_override=scope if scope else None,  # type: ignore[arg-type]
    )

    if not chunks:
        return "No extractable knowledge found in this transcript."

    # Save each chunk to the appropriate store
    saved = 0
    for chunk in chunks:
        try:
            meta = MetadataStore(chunk.scope, project_root or None)
            vector = VectorStore(chunk.scope, project_root or None)

            inserted = meta.insert(chunk)
            vector.index(inserted.id, inserted.content, inserted.tags)

            meta.close()
            vector.close()
            saved += 1
        except Exception as err:
            print(f"Failed to save chunk: {err}", file=sys.stderr)

    summary = "\n".join(
        f"- [{c.type}] {c.content[:80]}..." for c in chunks
    )

    return f"Extracted {len(chunks)} knowledge chunks, saved {saved}.\n\n{summary}"

File: dart/lib/tools/learn.dart
import 'dart:io';
import 'package:mcp_dart/mcp_dart.dart';
import '../extractor/extractor.dart';
import '../store/metadata.dart';
import '../store/vector.dart';
import '../store/scope.dart';
import '../store/types.dart';

void registerLearnTool(McpServer server) {
  server.registerTool(
    'learn',
    description: 'Extract and save knowledge from a conversation transcript',
    inputSchema: ToolInputSchema(
      properties: {
        'transcript_path': JsonSchema.string(
          description: 'Path to the .jsonl transcript file',
        ),
        'session_id': JsonSchema.string(
          description: 'Session ID for tracking source',
        ),
        'scope': JsonSchema.string(
          enumValues: ['global', 'project'],
          description: 'Force scope (default: auto-detect per chunk)',
        ),
      },
      required: ['transcript_path', 'session_id'],
    ),
    callback: (args, extra) async {
      final transcriptPath = args['transcript_path'] as String;
      final sessionId = args['session_id'] as String;
      final scopeStr = args['scope'] as String?;

      final projectRoot = detectProjectRoot();
      final projectName = projectRoot?.split('/').last;

      final scopeOverride =
          scopeStr != null ? KnowledgeScope.fromString(scopeStr) : null;

      // Extract knowledge from transcript
      final chunks = await extractKnowledge(
        transcriptPath: transcriptPath,
        sessionId: sessionId,
        trigger: ExtractionTrigger.manual,
        projectName: projectName,
        scopeOverride: scopeOverride,
      );

      if (chunks.isEmpty) {
        return CallToolResult(
          content: [
            TextContent(
              text: 'No extractable knowledge found in this transcript.',
            ),
          ],
        );
      }

      // Save each chunk to the appropriate store
      var saved = 0;
      for (final chunk in chunks) {
        try {
          final meta = MetadataStore(chunk.scope, projectRoot: projectRoot);
          final vector = VectorStore(chunk.scope, projectRoot: projectRoot);

          final inserted = meta.insert(chunk);
          vector.index(inserted.id, inserted.content, inserted.tags);

          meta.close();
          vector.close();
          saved++;
        } catch (err) {
          stderr.writeln('Failed to save chunk: $err');
        }
      }

      final summary =
          chunks.map((c) => '- [${c.type.name}] ${c.content.length > 80 ? '${c.content.substring(0, 80)}...' : c.content}').join('\n');

      return CallToolResult(
        content: [
          TextContent(
            text: 'Extracted ${chunks.length} knowledge chunks, saved $saved.\n\n$summary',
          ),
        ],
      );
    },
  );
}
